OPAL/ll/src/main/scala/org/opalj/ll/llvm/value/Instruction.scala

```diff
@@ -0,0 +1,335 @@
+
+    override def toString: String = {
+        s"${this.getClass.getSimpleName}(${repr})"
+    }
+}
+
+sealed abstract class BinaryOperation(ref: LLVMValueRef) extends Instruction(ref) {
+    def op1: Value = operand(0)
+    def op2: Value = operand(1)
+}
+
+sealed abstract class ConversionOperation(ref: LLVMValueRef) extends Instruction(ref) {
+    def value: Value = operand(0) // value
+    def newType: Value = operand(1) // type
+}
+
+case class Ret(ref: LLVMValueRef) extends Instruction(ref) with Terminator {
+    def value: Option[Value] = if (numOperands == 0 /* return void */ ) None else Some(operand(0))
+}
+case class Br(ref: LLVMValueRef) extends Instruction(ref) with Terminator
+case class Switch(ref: LLVMValueRef) extends Instruction(ref) with Terminator
+case class IndirectBr(ref: LLVMValueRef) extends Instruction(ref) with Terminator
+case class Invoke(ref: LLVMValueRef) extends Instruction(ref) with Terminator
+case class Unreachable(ref: LLVMValueRef) extends Instruction(ref) with Terminator
+case class CallBr(ref: LLVMValueRef) extends Instruction(ref) with Terminator
+case class FNeg(ref: LLVMValueRef) extends Instruction(ref)
+case class Add(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class FAdd(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class Sub(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class FSub(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class Mul(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class FMul(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class UDiv(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class SDiv(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class FDiv(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class URem(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class SRem(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class FRem(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class Shl(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class LShr(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class AShr(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class And(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class Or(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class Xor(ref: LLVMValueRef) extends BinaryOperation(ref)
+case class Alloca(ref: LLVMValueRef) extends Instruction(ref) {
+    def allocatedType: Type = Type(LLVMGetAllocatedType(ref))
+}
+case class Load(ref: LLVMValueRef) extends Instruction(ref) {
+    def src: Value = operand(0)
+}
+case class Store(ref: LLVMValueRef) extends Instruction(ref) {
+    def src: Value = operand(0)
+    def dst: Value = operand(1)
+}
+case class GetElementPtr(ref: LLVMValueRef) extends Instruction(ref) {
+    def base: Value = operand(0)
+    def isConstant = (1 until numOperands).forall(operand(_).isInstanceOf[ConstantIntValue])
+    def constants = (1 until numOperands).map(operand(_).asInstanceOf[ConstantIntValue].signExtendedValue)
```

Comment:
This can lead to an exception, right? At least document the conditions under which it is legal to invoke the method